### PR TITLE
feat: sign in with token endpoint

### PR DESCRIPTION
**Pull request by Timmy-id**
## What does the PR do?

This PR is about requesting and using an OTP to log in. The endpoint returns errors if:

- [x] User is not registered.
- [x] User inputs the wrong OTP.
- [x] Validation errors on the user inputs invalid or wrong details

How should this be manually tested?

1. Enter your email to request for OTP

2. Make a POST request to the `\api\v1\auth\signin-token` endpoint with the following payload:

```json
{
  "email": "tim@example.com",
}
```

3. A 6-digit OTP is then sent to the user's registered email

4. Then, make a POST request to the `\api\v1\auth\verify-signin-token` endpoint with the following payload:

```json
{
  "email": "tim@example.com",
  "token": "123456"
}
```

Successful Response

On a successful testimonial creation, the API should return a 200 OK status code.

```json
    {
      message: 'Sign-in token sent to email',
      status_code: 200,
    }
```

```json
    {
      message: 'Sign-in successful',
      "token": "eyJhbGciOiJI.......",
      "status_code": 200
    }
```

Authentication Error:

When the user is not registered:

```json
{
  "message": "User not found",
  "status_code": 404
}
```

Invalid OTP error

On Invalid OTP error, if the user is putting an invalid OTP or email:

```json
{
  "message": "Invalid token or email",
  "status_code": 401
}
```

Validation:

The response should return a 400 Bad Request status code with appropriate validation error messages if any required field is missing or invalid.

```json
{
  "message": ["email must be a string", "email must be an email"],
  "status_code": "400"
}
```

Task

- [x] Create a POST endpoint POST /api/v1/auth/signin-token/ to request for OTP for registered users.
- [x] Implement server-side logic to handle successful and failed update requests.
- [x] Ensure the registered user receives an OTP in their email.
- [x] The OTP is properly saved in the database by hashing it
- [x] Check that the OTP is valid
- [x] Then the user gets a token after all validations are passed

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/140)

### CODE REVIEW

**Comment by davidolanrewaju** [comment]
Change your title and PR template to follow the approved standard

**Comment by davidolanrewaju** [comment]
Also fix your workflow

**Comment by Timmy-id** [comment]
> Also fix your workflow

Kindly check the updated PR @davidolanrewaju

File: src/modules/auth/auth.controller.ts
Comment (Heba-WebDev) [inline src/modules/auth/auth.controller.ts:1]: response.status(HttpStatus.OK) is not needed in nestjs, change this please

**Review by Heba-WebDev — CHANGES_REQUESTED** [review]
Left you a comment, please take a look

File: src/modules/auth/auth.controller.ts
Comment (Timmy-id) [reply src/modules/auth/auth.controller.ts:1]: I've fixed it.

File: package.json [hunk 1]
```diff
@@ -43,7 +43,6 @@
     "@nestjs/swagger": "^7.3.1",
     "@nestjs/typeorm": "^10.0.2",
     "@types/nodemailer": "^6.4.15",
-    "bcrypt": "^5.1.1",
```
Comment (Homoakin619) [inline package.json:46]: why did you remove the package, return it. You shouldn't tamper with configurations

File: package.json [hunk 2]
```diff
@@ -66,7 +65,7 @@
     "@nestjs/cli": "^10.0.0",
     "@nestjs/schematics": "^10.0.0",
     "@nestjs/testing": "^10.0.0",
-    "@types/bcrypt": "^5.0.2",
```
Comment (Homoakin619) [inline package.json:69]: same as this, revert this change

File: src/modules/auth/auth.service.ts [hunk 3]
```diff
@@ -1,5 +1,12 @@
-import { HttpException, HttpStatus, Injectable, Logger } from '@nestjs/common';
-import * as bcrypt from 'bcrypt';
```
Comment (Homoakin619) [inline src/modules/auth/auth.service.ts:2]: why did you remove the package

File: src/modules/auth/auth.service.ts
Comment (Homoakin619) [inline src/modules/auth/auth.service.ts:1]: why did you tamper with someone else's code ?

File: src/modules/auth/auth.service.ts [hunk 4]
```diff
@@ -100,8 +111,12 @@ export default class AuthenticationService {
         };
       }
 
-      const token = (await this.otpService.createOtp(user.id)).token;
-      await this.emailService.sendForgotPasswordMail(dto.email, `${process.env.BASE_URL}/auth/reset-password`, token);
+      const newOtp = generateOtp(6);
+      const salt = await bcrypt.genSalt(10);
```
Comment (Homoakin619) [inline src/modules/auth/auth.service.ts:115]: why did you tamper with someone's code, revert the changes

File: src/modules/auth/auth.service.ts (hunk 3 above)
Comment (Homoakin619) [inline src/modules/auth/auth.service.ts:2]: why did you tamper with this, revert it

File: src/modules/user/user.service.ts [hunk 5]
```diff
@@ -39,7 +39,7 @@ export default class UserService {
     await this.userRepository.save(user);
   }
 
-  private async getUserByEmail(email: string) {
```
Comment (Homoakin619) [inline src/modules/user/user.service.ts:42]: why did you choose to tamper with codes

**Comment by Timmy-id** [comment]
Fixed @Homoakin619

**Comment by Timmy-id** [comment]
Fixed @Homoakin619

**Pull request closed without merging**